marconi/src/Marconi/Index/Utxo.hs

```diff
@@ -2,84 +2,193 @@
+| Address | TxId | TxIdx | DatumHash | Datum | Value | InlineScriptHash | InlineScript | Slot | BlockNumber |
+|---------+------+-------+-----------+-------+-------+------------------+--------------+------+-------------|
+
++ table: spent
+  |------+------|
+  | txId | txIx |
+  |------+------|
+
+-}
+module Marconi.Index.Utxo
+    ( eventAtAddress
+    , Utxo (..)
+    , utxoAddress
+    , utxoEventSlotNo
+    , UtxoEvent (..)
+    , utxoEventUtxos
+    , UtxoRow (..)
+    , utxoRowUtxo
+    , C.BlockNo (..)
+    , C.SlotNo (..)
+    , Depth (..)
+    , Result
+    , toRows
+    , addressFilteredRows
+    , toAddr
+    , UtxoIndex
+    , open
+    , query
+    , queryPlusVolatile
+    ) where
+
+import Codec.Serialise (Serialise (encode), deserialiseOrFail, serialise)
+import Codec.Serialise.Class (Serialise (decode))
+import Control.Concurrent.Async (concurrently_)
+import Control.Exception (bracket_)
+import Control.Lens (filtered, folded, traversed)
+import Control.Lens.Operators ((%~), (&), (^.), (^..))
+import Control.Lens.TH (makeLenses)
+import Control.Monad (unless, when)
+import Data.Aeson (ToJSON (toJSON))
+import Data.Aeson qualified
+import Data.ByteString.Lazy (toStrict)
+import Data.List (union)
 import Data.Maybe (fromJust)
 import Data.Proxy (Proxy (Proxy))
-import Data.Set (Set)
 import Data.Set qualified as Set
-import Database.SQLite.Simple (Only (Only), SQLData (SQLBlob, SQLInteger))
+import Data.Text (pack)
+import Database.SQLite.Simple (Only (Only), SQLData (SQLBlob, SQLInteger, SQLText))
 import Database.SQLite.Simple qualified as SQL
 import Database.SQLite.Simple.FromField (FromField (fromField), ResultError (ConversionFailed), returnError)
 import Database.SQLite.Simple.FromRow (FromRow (fromRow), field)
 import Database.SQLite.Simple.ToField (ToField (toField))
 import Database.SQLite.Simple.ToRow (ToRow (toRow))
 import GHC.Generics (Generic)
 import System.Random.MWC (createSystemRandom, uniformR)
+import Text.ParserCombinators.Parsec (parse)
 
+import Cardano.Api qualified as C
+import Cardano.Binary (fromCBOR, toCBOR)
+import Marconi.Types (CurrentEra)
 import RewindableIndex.Index.VSqlite (SqliteIndex)
 import RewindableIndex.Index.VSqlite qualified as Ix
 
-import Marconi.Types (CurrentEra, TxOut, TxOutRef, txOutRef)
+data Utxo               = Utxo
+    { _utxoAddress   :: !C.AddressAny
+    , _utxoTxId      :: !C.TxId
+    , _utxoTxIx      :: !C.TxIx
+    , _utxoDatum     :: Maybe C.ScriptData
+    , _utxoDatumHash :: Maybe (C.Hash C.ScriptData)
+    , _utxoValue     :: C.Value
+    -- , _inlineScriptHash    :: Maybe (C.Hash (C.ScriptDatum C.WitCtxTxIn)) --TODO
+    -- , _inlineScript        :: Maybe (C.ScriptDatum C.WitCtxTxIn) --TODO
+    } deriving (Show, Generic)
 
-data UtxoUpdate = UtxoUpdate
-  { _inputs  :: !(Set TxIn)
-  , _outputs :: ![(TxOut, TxOutRef)]
-  , _slotNo  :: !SlotNo
-  } deriving (Show)
+$(makeLenses ''Utxo)
 
-$(makeLenses ''UtxoUpdate)
+instance ToJSON C.AddressAny where
```

Comment:
Perhaps move the ToJSON instances in Marconi.Index.Utxo to where they are used. They are orphan instances anyway and may cause trouble when within a library.